### PR TITLE
List patterns: use the same binding path as range indexers

### PR DESCRIPTION
**Pull request by alrz**
Test plan: https://github.com/dotnet/roslyn/issues/51289

### CODE REVIEW

**Comment by alrz** [comment]
Technically RFR. Needs a rebase after previous PR merged in.

File: src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs [hunk 1]
```diff
@@ -186,7 +186,8 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Bindin
             bool permitDesignations,
             ref bool hasErrors,
             bool misplaced,
-            BindingDiagnosticBag diagnostics)
+            BindingDiagnosticBag diagnostics,
+            bool reportUnsupported)
```
Comment (alrz) [inline src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:190]: This is needed to report unsupported type for slice if we don't have a length property.
In that case we already reported unsupported type for list so we could probably live with that. However, we will have to adjust the tests to account for this which wouldn't be exactly per spec.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 2]
```diff
@@ -7941,15 +7955,13 @@ private BoundExpression BindIndexedPropertyAccess(SyntaxNode syntax, BoundExpres
...
-            SyntaxNode syntax,
             LookupResult lookupResult,
             BoundExpression? receiverOpt,
             TypeSymbol receiverType,
             bool argIsIndex,
             [NotNullWhen(true)] out Symbol? patternSymbol,
             BindingDiagnosticBag diagnostics,
-            ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo,
-            bool forPattern = false)
+            ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7964]: Not for this PR: I was looking at this code some more (to merge conflicts from `main` branch) and wondered why we have both diagnostics and use-site info in this signature and some relates ones.

**Comment by alrz** [comment]
Merged to https://github.com/dotnet/roslyn/pull/54335

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 2 above)
Comment (alrz) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7966]: @jcouv Should we create a DiagnosticBag off of useSiteInfo where we need one? I don't know if that's possible though.

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7966]: @alrz I can't look at the full code at the moment, so this is a general thought:
If this method selects a property to use, then a diagnostic bag should be enough (no need to delay the "use"). 
If the method only returns a candidate, then it should only produce use-site information but not use a diagnostic bag. The caller which decides to use the candidate is responsible for recording the use (ie transferring use-site information into diagnostic bag).

Comment (alrz) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:7966]: Yeah I think we should use useSiteInfo for everything here e.g. obsolete errors and others.

I was looking for `AddDiagnostics(UseSiteInfo<TAssemblySymbol> info)` I think that's the one we should use.

**Pull request closed without merging**